base/src/main/java/org/aya/tyck/ExprTycker.java

```diff
@@ -388,28 +395,65 @@ else if (iter.hasNext()) {
       }
       case Expr.LamExpr lam -> {
         if (term instanceof CallTerm.Hole) unifyTy(term, generatePi(lam), lam.sourcePos());
-        if (!(term.normalize(state, NormalizeMode.WHNF) instanceof FormTerm.Pi dt)) {
-          yield fail(lam, term, BadTypeError.pi(state, lam, term));
-        }
-        var param = lam.param();
-        if (param.explicit() != dt.param().explicit()) {
-          yield fail(lam, dt, new LicitProblem.LicitMismatchError(lam, dt));
-        }
-        var var = param.ref();
-        var lamParam = param.type();
-        var type = dt.param().type();
-        var result = synthesize(lamParam).wellTyped;
-        var comparison = unifyTy(result, type, lamParam.sourcePos());
-        if (comparison != null) {
-          // TODO: maybe also report this unification failure?
-          yield fail(lam, dt, BadTypeError.lamParam(state, lam, type, result));
-        } else type = result;
-        var resultParam = new Term.Param(var, type, param.explicit());
-        var body = dt.substBody(resultParam.toTerm());
-        yield localCtx.with(resultParam, () -> {
-          var rec = inherit(lam.body(), body);
-          return new Result(new IntroTerm.Lambda(resultParam, rec.wellTyped), dt);
-        });
+        yield switch (term.normalize(state, NormalizeMode.WHNF)) {
+          case FormTerm.Pi dt -> {
+            var param = lam.param();
+            if (param.explicit() != dt.param().explicit()) {
+              yield fail(lam, dt, new LicitProblem.LicitMismatchError(lam, dt));
+            }
+            var var = param.ref();
+            var lamParam = param.type();
+            var type = dt.param().type();
+            var result = synthesize(lamParam).wellTyped;
+            var comparison = unifyTy(result, type, lamParam.sourcePos());
+            if (comparison != null) {
+              // TODO: maybe also report this unification failure?
+              yield fail(lam, dt, BadTypeError.lamParam(state, lam, type, result));
+            } else type = result;
+            var resultParam = new Term.Param(var, type, param.explicit());
+            var body = dt.substBody(resultParam.toTerm());
+            yield localCtx.with(resultParam, () -> {
+              var rec = inherit(lam.body(), body);
+              return new Result(new IntroTerm.Lambda(resultParam, rec.wellTyped), dt);
+            });
+          }
+          // Path lambda!
+          case FormTerm.Path path -> {
+            var cubeParams = path.cube().params();
+            var plam = Expr.unPathLam(lam, cubeParams.size());
```

Comment:
Beware that there is a minor thing we can improve from guest: if the term is not a lam but can be eta-expand to a lam, then this code will fail